src/lib/battery/battery_base.cpp

```diff
@@ -0,0 +1,254 @@
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file battery_base.cpp
+ *
+ * Library calls for battery functionality.
+ *
+ * @author Julian Oes <julian@oes.ch>
+ * @author Timothy Scott <timothy@auterion.com>
+ */
+
+#include "battery_base.h"
+#include <mathlib/mathlib.h>
+#include <cstring>
+#include <px4_defines.h>
+
+BatteryBase::BatteryBase() :
+	ModuleParams(nullptr),
+	_warning(battery_status_s::BATTERY_WARNING_NONE),
+	_last_timestamp(0)
+{
+}
+
+void
+BatteryBase::reset()
+{
+	memset(&_battery_status, 0, sizeof(_battery_status));
+	_battery_status.current_a = -1.f;
+	_battery_status.remaining = 1.f;
+	_battery_status.scale = 1.f;
+	_battery_status.cell_count = _get_bat_n_cells();
+	// TODO: check if it is sane to reset warning to NONE
+	_battery_status.warning = battery_status_s::BATTERY_WARNING_NONE;
+	_battery_status.connected = false;
+	_battery_status.capacity = _get_bat_capacity();
+
+}
+
+void
+BatteryBase::updateBatteryStatus(hrt_abstime timestamp, float voltage_v, float current_a, bool connected,
+				 bool selected_source, int priority, float throttle_normalized, bool armed, bool shouldPublish)
+{
+	reset();
+	_battery_status.timestamp = timestamp;
+	filterVoltage(voltage_v);
+	filterThrottle(throttle_normalized);
+	filterCurrent(current_a);
+	sumDischarged(timestamp, current_a);
+	estimateRemaining(_voltage_filtered_v, _current_filtered_a, _throttle_filtered, armed);
+	computeScale();
+
+	if (_battery_initialized) {
+		determineWarning(connected);
+	}
+
+	if (_voltage_filtered_v > 2.1f) {
+		_battery_initialized = true;
+		_battery_status.voltage_v = voltage_v;
+		_battery_status.voltage_filtered_v = _voltage_filtered_v;
+		_battery_status.scale = _scale;
+		_battery_status.current_a = current_a;
+		_battery_status.current_filtered_a = _current_filtered_a;
+		_battery_status.discharged_mah = _discharged_mah;
+		_battery_status.warning = _warning;
+		_battery_status.remaining = _remaining;
+		_battery_status.connected = connected;
+		_battery_status.system_source = selected_source;
+		_battery_status.priority = priority;
+	}
+
+	_battery_status.timestamp = timestamp;
+
+	if (shouldPublish) {
+		publish();
+	}
+
+	_battery_status.temperature = NAN;
+}
+
+void
+BatteryBase::publish()
+{
+	orb_publish_auto(ORB_ID(battery_status), &_orbAdvert, &_battery_status, &_orbInstance, ORB_PRIO_DEFAULT);
+}
+
+void
+BatteryBase::filterVoltage(float voltage_v)
+{
+	if (!_battery_initialized) {
+		_voltage_filtered_v = voltage_v;
+	}
+
+	// TODO: inspect that filter performance
+	const float filtered_next = _voltage_filtered_v * 0.99f + voltage_v * 0.01f;
+
+	if (PX4_ISFINITE(filtered_next)) {
```

Comment:
Why would this ever not be finite?

Reply:
If voltage_v is for some reason NaN, I guess. I did not write this part, it was in the original battery library, so I assumed there was a reason for it.